File: packages/external/agent/src/capability-fusion.ts
/**
 * @flowforge/external-agent capability-fusion — F35 能力融合（EX-010）。
 *
 * TS 重写自 flowforge/core/external_agent/capability_fusion.py：
 *   - FusionConfig: base_weight=0.1 / max_weight=0.5 / min_invocations=3 /
 *     min_success_rate=0.7 / fuse_blind_spots=true
 *   - FusionResult: fused_profile / fused / fused_capabilities /
 *     fused_blind_spots / fusion_weight / reason
 *   - fuse(forgekinProfile, externalAgentProfile, invocationCount,
 *     successRate)：门槛检查 → weight = min(base*count, max) →
 *     能力不去重合并 → 盲点合并 → 历史追加
 */

/** 融合配置（capability_fusion.py FusionConfig）。 */
export interface FusionConfig {
  /** 单次调用的基础权重。 */
  readonly base_weight: number;
  /** 权重上限。 */
  readonly max_weight: number;
  /** 融合最低调用次数门槛。 */
  readonly min_invocations: number;
  /** 融合最低成功率门槛。 */
  readonly min_success_rate: number;
  /** 是否融合盲点（EX-002）。 */
  readonly fuse_blind_spots: boolean;
}

/** 默认融合配置（capability_fusion.py 缺省值）。 */
export const DEFAULT_FUSION_CONFIG: FusionConfig = {
  base_weight: 0.1,
  max_weight: 0.5,
  min_invocations: 3,
  min_success_rate: 0.7,
  fuse_blind_spots: true,
};

/** 融合结果（capability_fusion.py FusionResult）。 */
export interface FusionResult {
  /** 融合后的 Forgekin 画像（原画像 + 新能力）。 */
  readonly fused_profile: Record<string, unknown>;
  /** 是否发生了融合（门槛未过时为 false）。 */
  readonly fused: boolean;
  /** 新增能力列表（未去重合并）。 */
  readonly fused_capabilities: readonly string[];
  /** 新增盲点列表。 */
  readonly fused_blind_spots: readonly string[];
  /** 融合权重（base*count 封顶 max）。 */
  readonly fusion_weight: number;
  /** 未融合的原因（fused=false 时）。 */
  readonly reason: string;
}

/** 能力融合历史条目。 */
export interface FusionHistoryEntry {
  readonly provider_name: string;
  readonly fused_capabilities: readonly string[];
  readonly fused_blind_spots: readonly string[];
  readonly fusion_weight: number;
  readonly timestamp: string;
}

/** 外部 Agent 能力融合（capability_fusion.py ExternalAgentCapabilityFusion）。 */
export class ExternalAgentCapabilityFusion {
  private readonly _config: FusionConfig;
  /** 融合历史（provider_name -> entries）。 */
  private readonly _history = new Map<string, FusionHistoryEntry[]>();

  constructor(config: Partial<FusionConfig> = {}) {
    this._config = { ...DEFAULT_FUSION_CONFIG, ...config };
  }

  /**
   * 融合外部 Agent 能力到 Forgekin 画像（EX-010）。
   *
   * 门槛检查（capability_fusion.py fuse）：
   *   - invocation_count < min_invocations → 不融合
   *   - success_rate < min_success_rate → 不融合
   * 权重：weight = min(base_weight * invocation_count, max_weight)。
   */
  fuse(
    forgekinProfile: Record<string, unknown>,
    externalAgentProfile: Record<string, unknown>,
    invocationCount: number,
    successRate: number,
  ): FusionResult {
    // 1. 门槛检查
    if (invocationCount < this._config.min_invocations) {
      return {
        fused_profile: forgekinProfile,
        fused: false,
        fused_capabilities: [],
        fused_blind_spots: [],
        fusion_weight: 0,
        reason:
          `invocation_count=${invocationCount} < ` +
          `min_invocations=${this._config.min_invocations}`,
      };
    }
    if (successRate < this._config.min_success_rate) {
      return {
        fused_profile: forgekinProfile,
        fused: false,
        fused_capabilities: [],
        fused_blind_spots: [],
        fusion_weight: 0,
        reason:
          `success_rate=${successRate} < ` +
          `min_success_rate=${this._config.min_success_rate}`,
      };
    }

    // 2. 权重计算
    const weight = Math.min(
      this._config.base_weight * invocationCount,
      this._config.max_weight,
    );

    // 3. 能力不去重合并
    const externalCapabilities = asStrArray(externalAgentProfile.capabilities);
    const profileCapabilities = asStrArray(forgekinProfile.capabilities);
    const fusedCapabilities = [...profileCapabilities, ...externalCapabilities];

    // 4. 盲点合并（forgekin + external，不去重）
    const profileBlindSpots = asStrArray(forgekinProfile.blind_spots);
    const fusedBlindSpots = this._config.fuse_blind_spots
      ? [...profileBlindSpots, ...asStrArray(externalAgentProfile.blind_spots)]
      : [];

    // 5. 融合画像
    const fusedProfile: Record<string, unknown> = {
      ...forgekinProfile,
      capabilities: fusedCapabilities,
    };
    if (this._config.fuse_blind_spots) {
      fusedProfile.blind_spots = [...fusedBlindSpots];
    }

    // 6. 追加历史
    const providerName =
      typeof externalAgentProfile.provider_name === 'string'
        ? (externalAgentProfile.provider_name as string)
        : 'unknown';
    const entry: FusionHistoryEntry = {
      provider_name: providerName,
      fused_capabilities: externalCapabilities,
      fused_blind_spots: fusedBlindSpots,
      fusion_weight: weight,
      timestamp: new Date().toISOString(),
    };
    const history = this._history.get(providerName) ?? [];
    history.push(entry);
    this._history.set(providerName, history);

    return {
      fused_profile: fusedProfile,
      fused: true,
      fused_capabilities: fusedCapabilities,
      fused_blind_spots: fusedBlindSpots,
      fusion_weight: weight,
      reason: 'fused',
    };
  }

  /** 查询某 Provider 的融合历史（按时间升序）。 */
  getFusionHistory(providerName: string): FusionHistoryEntry[] {
    return [...(this._history.get(providerName) ?? [])];
  }

  /** 当前融合配置。 */
  get config(): FusionConfig {
    return { ...this._config };
  }
}

function asStrArray(value: unknown): string[] {
  if (!Array.isArray(value)) {
    return [];
  }
  return value.map((item) => String(item));
}

File: python/legacy/core/external_agent/capability_fusion.py
"""ExternalAgentCapabilityFusion — 三方 Agent 能力融合机制（EX-010）。

Forgekin调用三方 Agent 后，三方 Agent 的能力应能"沉淀"到Forgekin的能力画像中。
如Forgekin多次调用 claude code 写代码后，应"学到"代码编写能力（通过SpiritForge蒸馏）。

设计依据：
    - [doc:review/review.md#第九章§9.2] EX-010 三方 Agent 与Forgekin能力融合机制缺失
    - [doc:decisions/006-external-agent-integration.md] §3 四大机制（F035）
    - [doc:design/naming-contract.md#2.7] SpiritForge / [doc:design/naming-contract.md#2.12] 能力画像

铁律遵守：
    - 铁律 3：依赖通过构造函数注入
    - 编程红线 9：使用组合表达融合策略
    - 不修改不相关代码（编程红线 7）

License: MIT
"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

from flowforge.core.tracing import get_logger

logger = get_logger("external_agent.capability_fusion")


class FusionConfig(BaseModel):
    """能力融合配置（控制融合强度 / 阈值 / 衰减）。"""

    # 融合权重：外部能力对Forgekin主画像的影响系数（0.0-1.0）
    # 调用次数越多、成功率越高，融合权重越大（但不超过 max_weight）
    base_weight: float = Field(default=0.1, ge=0.0, le=1.0, description="基础融合权重")
    max_weight: float = Field(default=0.5, ge=0.0, le=1.0, description="最大融合权重")
    # 触发融合的最小调用次数（避免一次调用就"学到"能力）
    min_invocations: int = Field(default=3, ge=1, description="触发融合的最小调用次数")
    # 触发融合的最小成功率（避免从失败中"学到"错误能力）
    min_success_rate: float = Field(
        default=0.7, ge=0.0, le=1.0, description="触发融合的最小成功率"
    )
    # 盲点融合开关：是否同时融合三方 Agent 的盲点（用于跨厂商 review 配对）
    fuse_blind_spots: bool = Field(default=True, description="是否融合盲点")


class FusionResult(BaseModel):
    """能力融合结果。"""

    fused_profile: dict[str, Any] = Field(..., description="融合后的能力画像")
    fused: bool = Field(..., description="是否实际发生融合")
    fused_capabilities: list[str] = Field(
        default_factory=list, description="本次融合的能力列表"
    )
    fused_blind_spots: list[str] = Field(
        default_factory=list, description="本次融合的盲点列表"
    )
    fusion_weight: float = Field(default=0.0, description="本次融合的实际权重")
    reason: str = Field(default="", description="未融合原因（fused=False 时）")


class ExternalAgentCapabilityFusion:
    """三方 Agent 能力融合机制（EX-010）。

    Forgekin调用三方 Agent 后，三方 Agent 的能力应能"沉淀"到Forgekin的能力画像中。
    如Forgekin多次调用 claude code 写代码后，应"学到"代码编写能力（通过SpiritForge蒸馏）。

    详见 [doc:review/review.md#第九章§9.2] EX-010

    融合策略：
        1. 调用次数门槛：min_invocations 次以下不融合（避免一次调用"学到"）
        2. 成功率门槛：min_success_rate 以下不融合（避免从失败中学习）
        3. 权重渐进：weight = min(base * invocation_count, max_weight)
        4. 盲点同步：盲点同步到Forgekin画像，用于跨厂商 review 配对
        5. 不去重：原有能力保留，新能力追加（让SpiritForge后续蒸馏去重）
    """

    def __init__(self, config: FusionConfig | None = None) -> None:
        """注入融合配置。

        Args:
            config: 融合配置（None 时使用默认配置）。
        """
        self._config = config or FusionConfig()

    def fuse(
        self,
        forgekin_profile: dict[str, Any],
        external_agent_profile: dict[str, Any],
        invocation_count: int,
        success_rate: float,
    ) -> FusionResult:
        """融合三方 Agent 能力到Forgekin主画像。

        Args:
            forgekin_profile: Forgekin当前能力画像（含 capabilities / blind_spots）。
            external_agent_profile: 三方 Agent 能力画像（来自 Adapter.get_capability_profile）。
            invocation_count: 历史调用次数（用于门槛判断）。
            success_rate: 历史成功率（0.0-1.0）。

        Returns:
            FusionResult 融合结果（fused_profile 是融合后的新画像，调用方负责持久化）。
        """
        # 门槛检查
        if invocation_count < self._config.min_invocations:
            return FusionResult(
                fused_profile=forgekin_profile,
                fused=False,
                reason=(
                    f"invocation_count={invocation_count} < "
                    f"min={self._config.min_invocations}"
                ),
            )
        if success_rate < self._config.min_success_rate:
            return FusionResult(
                fused_profile=forgekin_profile,
                fused=False,
                reason=(
                    f"success_rate={success_rate:.2f} < "
                    f"min={self._config.min_success_rate:.2f}"
                ),
            )

        # 计算融合权重（渐进式，不超过 max_weight）
        weight = min(
            self._config.base_weight * invocation_count,
            self._config.max_weight,
        )

        # 融合能力（不去重，让SpiritForge后续蒸馏去重）
        existing_caps: list[str] = list(forgekin_profile.get("capabilities", []))
        external_caps: list[str] = list(
            external_agent_profile.get("capabilities", [])
        )
        new_caps = [c for c in external_caps if c not in existing_caps]
        fused_caps = existing_caps + new_caps

        # 融合盲点（用于跨厂商 review 配对）
        fused_blind_spots: list[str] = list(
            forgekin_profile.get("blind_spots", [])
        )
        new_blind_spots: list[str] = []
        if self._config.fuse_blind_spots:
            external_blind = list(external_agent_profile.get("blind_spots", []))
            new_blind_spots = [
                b for b in external_blind if b not in fused_blind_spots
            ]
            fused_blind_spots = fused_blind_spots + new_blind_spots

        # 构建融合后的画像（保留原画像其他字段）
        fused_profile: dict[str, Any] = {
            **forgekin_profile,
            "capabilities": fused_caps,
            "blind_spots": fused_blind_spots,
            # 记录融合历史（供SpiritForge SpiritForge 蒸馏使用）
            "fusion_history": list(forgekin_profile.get("fusion_history", [])) + [
                {
                    "external_agent": external_agent_profile.get(
                        "provider_name", "unknown"
                    ),
                    "fused_capabilities": new_caps,
                    "fused_blind_spots": new_blind_spots,
                    "weight": weight,
                    "invocation_count": invocation_count,
                    "success_rate": success_rate,
                }
            ],
        }

        logger.info(
            "fusion.fuse new_caps=%d new_blind_spots=%d weight=%.3f invocations=%d",
            len(new_caps),
            len(new_blind_spots),
            weight,
            invocation_count,
        )

        return FusionResult(
            fused_profile=fused_profile,
            fused=True,
            fused_capabilities=new_caps,
            fused_blind_spots=new_blind_spots,
            fusion_weight=weight,
        )

    def get_config(self) -> FusionConfig:
        """返回当前融合配置（供调试 / 测试使用）。"""
        return self._config
